File: tests/open_position.ts
import { expect } from "chai";
import { Keypair, Transaction } from "@solana/web3.js";
import {
  BUY_SOL,
  Env,
  LAMPORTS_PER_SOL,
  NOW,
  SELL_SOL,
  USDC,
  alignedExpiry,
  expectErr,
  expectOk,
  nextNonce,
} from "./helpers";

const FIXED = 250n * USDC; // 250 USDC per SOL
const EXPIRY = alignedExpiry(7);

describe("open_position", () => {
  let env: Env;
  beforeEach(async () => {
    env = new Env();
    await env.bootstrap();
  });

  const base = () => ({
    product: SELL_SOL,
    fixedPrice: FIXED,
    expiryTs: EXPIRY,
    amount: 10n * LAMPORTS_PER_SOL,
    // Both products pay the premium in USDC.
    yieldAmount: 12n * USDC,
    nonce: nextNonce(),
  });

  describe("guards", () => {
    it("rejects an expiry in the past", async () => {
      const { res } = await env.openPosition({
        ...base(),
        expiryTs: alignedExpiry(-1),
      });
      expectErr(res, "ExpiryInPast");
    });

    it("rejects an expiry that is not 08:00 UTC", async () => {
      const { res } = await env.openPosition({
        ...base(),
        expiryTs: EXPIRY + 60,
      });
      expectErr(res, "ExpiryNotAligned");
    });

    it("rejects a zero fixed price", async () => {
      const { res } = await env.openPosition({ ...base(), fixedPrice: 0n });
      expectErr(res, "InvalidFixedPrice");
    });

    it("rejects a zero amount", async () => {
      const { res } = await env.openPosition({ ...base(), amount: 0n });
      expectErr(res, "InvalidAmount");
    });

    it("rejects a zero yield", async () => {
      const { res } = await env.openPosition({ ...base(), yieldAmount: 0n });
      expectErr(res, "InvalidYield");
    });

    it("rejects user == market maker", async () => {
      const { res } = await env.openPosition({ ...base(), user: env.mm });
      expectErr(res, "SameCounterparty");
    });

    it("rejects a USDC mint that differs from GlobalConfig", async () => {
      const other = Keypair.generate();
      // Create a second mint and fund both parties on it so only the mint check fails.
      const saved = env.usdcMint;
      env.usdcMint = other;
      env.createUsdcMint();
      env.mintUsdc(env.mm.publicKey, 1000n * USDC);
      env.mintUsdc(env.user.publicKey, 1000n * USDC);
      env.usdcMint = saved;
      const { res } = await env.openPosition({
        ...base(),
        usdcMint: other.publicKey,
      });
      expectErr(res, "InvalidUsdcMint");
    });

    it("fails without the market maker signature", async () => {
      const { ix } = await env.openPositionIx(base());
      const tx = new Transaction();
      tx.recentBlockhash = env.svm.latestBlockhash();
      tx.feePayer = env.mm.publicKey;
      tx.add(ix);
      tx.partialSign(env.user);
      expect(tx.signatures.some((s) => s.signature === null)).to.be.true;
      let threw = false;
      try {
        env.sendRaw(tx);
      } catch {
        threw = true;
      }
      expect(threw, "unsigned MM slot must be rejected").to.be.true;
    });

    it("fails without the user signature", async () => {
      const { ix } = await env.openPositionIx(base());
      const tx = new Transaction();
      tx.recentBlockhash = env.svm.latestBlockhash();
      tx.feePayer = env.mm.publicKey;
      tx.add(ix);
      tx.partialSign(env.mm);
      let threw = false;
      try {
        env.sendRaw(tx);
      } catch {
        threw = true;
      }
      expect(threw, "unsigned user slot must be rejected").to.be.true;
    });

    it("fails when the market maker cannot cover its leg", async () => {
      const poorMm = Keypair.generate();
      env.svm.airdrop(poorMm.publicKey, 5n * LAMPORTS_PER_SOL);
      env.mintUsdc(poorMm.publicKey, 1n * USDC);
      const { res } = await env.openPosition({ ...base(), mm: poorMm });
      expectErr(res, "insufficient funds");
    });
  });

  describe("Sell SOL happy path", () => {
    it("locks user SOL, MM USDC and pays yield in USDC, with MM as fee payer", async () => {
      const p = base();
      const userSolBefore = env.sol(env.user.publicKey);
      const mmSolBefore = env.sol(env.mm.publicKey);
      const mmUsdcBefore = env.usdcBalance(env.mm.publicKey);
      const userUsdcBefore = env.usdcBalance(env.user.publicKey);

      const { res, position } = await env.openPosition(p);
      expectOk(res);

      const pos = await env.fetchPosition(position);
      expect(pos.user.equals(env.user.publicKey)).to.be.true;
      expect(pos.marketMaker.equals(env.mm.publicKey)).to.be.true;
      expect(pos.product).to.deep.equal({ sellSol: {} });
      expect(pos.token).to.deep.equal({ sol: {} });
      expect(BigInt(pos.userCollateral.toString())).to.equal(p.amount);
      // 10 SOL * 250 USDC = 2500 USDC
      expect(BigInt(pos.mmCollateral.toString())).to.equal(2500n * USDC);
      expect(BigInt(pos.yieldAmount.toString())).to.equal(p.yieldAmount);
      expect(pos.settled).to.be.false;

      // User: -amount SOL, +yield USDC, no fees.
      expect(env.sol(env.user.publicKey)).to.equal(userSolBefore - p.amount);
      expect(env.usdcBalance(env.user.publicKey)).to.equal(
        userUsdcBefore + p.yieldAmount
      );
      // MM paid the fee and rent for position + vault (+ nothing for the user ATA, which exists).
      expect(env.sol(env.mm.publicKey) < mmSolBefore).to.be.true;
      // MM USDC: the payment leg plus the yield.
      expect(env.usdcBalance(env.mm.publicKey)).to.equal(
        mmUsdcBefore - 2500n * USDC - p.yieldAmount
      );
      expect(env.tokenAccountBalance(env.vault(position))).to.equal(
        2500n * USDC
      );
      // Position holds rent + user collateral.
      const rent = env.svm.minimumBalanceForRentExemption(
        BigInt(env.svm.getAccount(position)!.data.length)
      );
      expect(env.sol(position)).to.equal(rent + p.amount);
    });

    it("creates the user's USDC account when missing", async () => {
      const freshUser = Keypair.generate();
      env.svm.airdrop(freshUser.publicKey, 100n * LAMPORTS_PER_SOL);
      expect(env.svm.getAccount(env.ata(freshUser.publicKey))).to.be.null;
      const { res } = await env.openPosition({ ...base(), user: freshUser });
      expectOk(res);
      // The account exists and holds the upfront yield.
      expect(env.usdcBalance(freshUser.publicKey)).to.equal(base().yieldAmount);
      expect(env.svm.getAccount(env.ata(freshUser.publicKey))).to.not.be.null;
    });
  });

  describe("Buy SOL happy path", () => {
    it("locks user USDC, MM SOL and pays yield in USDC", async () => {
      const p = {
        ...base(),
        product: BUY_SOL,
        amount: 2500n * USDC, // 2500 USDC buys 10 SOL at 250
        yieldAmount: 40n * USDC,
      };
      const userUsdcBefore = env.usdcBalance(env.user.publicKey);
      const mmUsdcBefore = env.usdcBalance(env.mm.publicKey);
      const userSolBefore = env.sol(env.user.publicKey);

      const { res, position } = await env.openPosition(p);
      expectOk(res);

      const pos = await env.fetchPosition(position);
      expect(pos.product).to.deep.equal({ buySol: {} });
      expect(pos.token).to.deep.equal({ usdc: {} });
      expect(BigInt(pos.userCollateral.toString())).to.equal(p.amount);
      expect(BigInt(pos.mmCollateral.toString())).to.equal(
        10n * LAMPORTS_PER_SOL
      );

      expect(env.usdcBalance(env.user.publicKey)).to.equal(
        userUsdcBefore - p.amount + p.yieldAmount
      );
      expect(env.usdcBalance(env.mm.publicKey)).to.equal(
        mmUsdcBefore - p.yieldAmount
      );
      expect(env.tokenAccountBalance(env.vault(position))).to.equal(p.amount);
      expect(env.sol(env.user.publicKey)).to.equal(userSolBefore);
      const rent = env.svm.minimumBalanceForRentExemption(
        BigInt(env.svm.getAccount(position)!.data.length)
      );
      expect(env.sol(position)).to.equal(rent + 10n * LAMPORTS_PER_SOL);
    });
  });

  describe("replay protection", () => {
    it("rejects the same params twice and accepts a fresh nonce", async () => {
      const p = base();
      expectOk((await env.openPosition(p)).res);
      expectErr((await env.openPosition(p)).res, "already in use");
      expectOk((await env.openPosition({ ...p, nonce: nextNonce() })).res);
    });

    it("rejects an identical signed transaction re-submitted", async () => {
      const { ix } = await env.openPositionIx(base());
      const tx = new Transaction();
      tx.recentBlockhash = env.svm.latestBlockhash();
      tx.feePayer = env.mm.publicKey;
      tx.add(ix);
      tx.sign(env.mm, env.user);
      expectOk(env.svm.sendTransaction(tx));
      const again = env.svm.sendTransaction(tx);
      expect(again.constructor.name).to.equal("FailedTransactionMetadata");
    });
  });

  it("keeps the dual-signed transaction well under the 1232-byte limit", async () => {
    const { ix } = await env.openPositionIx(base());
    const tx = new Transaction();
    tx.recentBlockhash = env.svm.latestBlockhash();
    tx.feePayer = env.mm.publicKey;
    tx.add(ix);
    tx.sign(env.mm, env.user);
    const size = tx.serialize().length;
    expect(size < 1232).to.be.true;
    expect(size < 800).to.be.true;
  });

  it("uses the on-chain clock for the expiry check", async () => {
    env.setTime(EXPIRY + 1);
    const { res } = await env.openPosition(base());
    expectErr(res, "ExpiryInPast");
    env.setTime(NOW);
  });
});

File: programs/breezepocket/src/instructions/open_position.rs
use anchor_lang::prelude::*;
use anchor_lang::system_program;
use anchor_spl::associated_token::AssociatedToken;
use anchor_spl::token::{self, Mint, Token as TokenProgram, TokenAccount};

use crate::errors::ErrorCode;
use crate::state::*;

#[derive(AnchorSerialize, AnchorDeserialize, Clone)]
pub struct OpenParams {
    pub product: Product,
    /// USDC base units per SOL.
    pub fixed_price: u64,
    /// Must be 08:00 UTC on a future date.
    pub expiry_ts: i64,
    /// User collateral in base units of the product's collateral token.
    pub amount: u64,
    /// Yield paid upfront to the user, in USDC base units for both products.
    pub yield_amount: u64,
    /// Random per-quote value; part of the PDA seed so a quote can only be used once.
    pub nonce: u64,
}

#[derive(Accounts)]
#[instruction(params: OpenParams)]
pub struct OpenPosition<'info> {
    /// Fee payer, rent payer, co-signer, and source of MM collateral and yield.
    #[account(mut)]
    pub market_maker: Signer<'info>,

    #[account(mut)]
    pub user: Signer<'info>,

    #[account(seeds = [CONFIG_SEED], bump = config.bump)]
    pub config: Account<'info, GlobalConfig>,

    #[account(
        init,
        payer = market_maker,
        space = 8 + PositionAccount::INIT_SPACE,
        seeds = [
            POSITION_SEED,
            user.key().as_ref(),
            market_maker.key().as_ref(),
            &params.fixed_price.to_le_bytes(),
            &params.expiry_ts.to_le_bytes(),
            &params.nonce.to_le_bytes(),
        ],
        bump,
    )]
    pub position: Account<'info, PositionAccount>,

    #[account(address = config.usdc_mint @ ErrorCode::InvalidUsdcMint)]
    pub usdc_mint: Account<'info, Mint>,

    /// Holds the USDC leg of the position until settlement.
    #[account(
        init,
        payer = market_maker,
        associated_token::mint = usdc_mint,
        associated_token::authority = position,
    )]
    pub position_usdc_vault: Account<'info, TokenAccount>,

    /// Created by the market maker if the user has no USDC account yet, so that
    /// settlement can always pay the user in USDC.
    #[account(
        init_if_needed,
        payer = market_maker,
        associated_token::mint = usdc_mint,
        associated_token::authority = user,
    )]
    pub user_usdc_ata: Account<'info, TokenAccount>,

    #[account(
        mut,
        associated_token::mint = usdc_mint,
        associated_token::authority = market_maker,
    )]
    pub mm_usdc_ata: Account<'info, TokenAccount>,

    pub token_program: Program<'info, TokenProgram>,
    pub associated_token_program: Program<'info, AssociatedToken>,
    pub system_program: Program<'info, System>,
}

pub fn handle_open_position(ctx: Context<OpenPosition>, params: OpenParams) -> Result<()> {
    let now = Clock::get()?.unix_timestamp;
    require!(params.expiry_ts > now, ErrorCode::ExpiryInPast);
    require!(
        is_aligned_expiry(params.expiry_ts),
        ErrorCode::ExpiryNotAligned
    );
    require!(params.fixed_price > 0, ErrorCode::InvalidFixedPrice);
    require!(params.amount > 0, ErrorCode::InvalidAmount);
    require!(params.yield_amount > 0, ErrorCode::InvalidYield);
    require_keys_neq!(
        ctx.accounts.user.key(),
        ctx.accounts.market_maker.key(),
        ErrorCode::SameCounterparty
    );

    let mm_collateral = mm_collateral_for(params.product, params.amount, params.fixed_price)
        .ok_or(ErrorCode::MathOverflow)?;
    require!(mm_collateral > 0, ErrorCode::InvalidAmount);

    let accounts = &ctx.accounts;
    match params.product {
        Product::SellSol => {
            // User SOL -> position PDA.
            system_program::transfer(
                CpiContext::new(
                    accounts.system_program.to_account_info(),
                    system_program::Transfer {
                        from: accounts.user.to_account_info(),
                        to: accounts.position.to_account_info(),
                    },
                ),
                params.amount,
            )?;
            // MM USDC (payment leg) -> vault.
            token::transfer(
                CpiContext::new(
                    accounts.token_program.to_account_info(),
                    token::Transfer {
                        from: accounts.mm_usdc_ata.to_account_info(),
                        to: accounts.position_usdc_vault.to_account_info(),
                        authority: accounts.market_maker.to_account_info(),
                    },
                ),
                mm_collateral,
            )?;
            // Yield in USDC: MM -> user, upfront. Both products pay the premium in USDC.
            token::transfer(
                CpiContext::new(
                    accounts.token_program.to_account_info(),
                    token::Transfer {
                        from: accounts.mm_usdc_ata.to_account_info(),
                        to: accounts.user_usdc_ata.to_account_info(),
                        authority: accounts.market_maker.to_account_info(),
                    },
                ),
                params.yield_amount,
            )?;
        }
        Product::BuySol => {
            // User USDC -> vault.
            token::transfer(
                CpiContext::new(
                    accounts.token_program.to_account_info(),
                    token::Transfer {
                        from: accounts.user_usdc_ata.to_account_info(),
                        to: accounts.position_usdc_vault.to_account_info(),
                        authority: accounts.user.to_account_info(),
                    },
                ),
                params.amount,
            )?;
            // MM SOL (delivery leg) -> position PDA.
            system_program::transfer(
                CpiContext::new(
                    accounts.system_program.to_account_info(),
                    system_program::Transfer {
                        from: accounts.market_maker.to_account_info(),
                        to: accounts.position.to_account_info(),
                    },
                ),
                mm_collateral,
            )?;
            // Yield in USDC: MM -> user, upfront.
            token::transfer(
                CpiContext::new(
                    accounts.token_program.to_account_info(),
                    token::Transfer {
                        from: accounts.mm_usdc_ata.to_account_info(),
                        to: accounts.user_usdc_ata.to_account_info(),
                        authority: accounts.market_maker.to_account_info(),
                    },
                ),
                params.yield_amount,
            )?;
        }
    }

    let position = &mut ctx.accounts.position;
    position.user = ctx.accounts.user.key();
    position.market_maker = ctx.accounts.market_maker.key();
    position.product = params.product;
    position.token = params.product.collateral_token();
    position.fixed_price = params.fixed_price;
    position.expiry_ts = params.expiry_ts;
    position.user_collateral = params.amount;
    position.mm_collateral = mm_collateral;
    position.yield_amount = params.yield_amount;
    position.nonce = params.nonce;
    position.settled = false;
    position.bump = ctx.bumps.position;
    Ok(())
}
